contracts/random/RandomSeedProviderRequestQueue.sol

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2023
+// SPDX-License-Identifier: Apache 2
+pragma solidity 0.8.19;
+
+/**
+ * @notice Queue to be used with RandomSeedProvider
+ * @dev    The contract is upgradeable.
+ */
+contract RandomSeedProviderRequestQueue {
+    // Error thrown if a block is requested that isn't in the queue.
+    error BlockNotFound(uint256 _blockNumber);
+
+    /// @notice Block numbers for which there have been requests to generate a seed value.
+    mapping(uint256 index => uint256 blockNumber) private outstandingRequests;
+
+    /// @notice Tail of the FIFO
+    uint256 private outstandingRequestsTail;
+
+    /// @notice Head of the FIFO
+    uint256 private outstandingRequestsHead;
+
+    function initializeRandomSeedProviderRequestQueue() internal {
+        outstandingRequestsTail = 0;
+        outstandingRequestsHead = 0;
+    }
+
+    /**
+     * @notice Enqueue a block number if it is unique.
+     * @param _blockNumber The number to add to the queue.
+     */
+    function enqueueIfUnique(uint256 _blockNumber) internal {
+        // If the queue is not empty, check the latest value put into the tail of the queue.
+        // If it is the same as the block number to be enqueued then there is nothing to do.
+        uint256 tail = outstandingRequestsTail;
+        uint256 head = outstandingRequestsHead;
+        if (head != tail) {
+            // slither-disable-next-line incorrect-equality
+            if (outstandingRequests[tail] == _blockNumber) {
```

Comment:
would this uniqueness check work if the onchain delay is updated (reduced between requests?